File: FluentMessageSender/MessageSender.vb
Namespace Fluent
    Public Class MessageSender
        Private ReadOnly _messageSender As IMessageSender

        Public Sub New (ByVal sender As IMessageSender)
            _messageSender = sender
        End Sub

        Public Function SendText (ByVal text As String) As SendExpression
            Return New SendExpression (text, _messageSender)
        End Function

        Public Class SendExpression
            Implements ToExpression

            Private ReadOnly _text As String
            Private ReadOnly _messageSender As IMessageSender
            Private _sender As String

            Public Sub New(ByVal text As String, ByVal messageSender As IMessageSender)
                _text = text
                _messageSender = messageSender
            End Sub

            Public Function From(ByVal sender As FromAddress) As ToExpression
                _sender = sender
                Return Me
            End Function

            Public Sub ToExpression(ByVal receiver As String) Implements ToExpression.To
                _messageSender.SendMessage(_text, _sender, receiver)
            End Sub
        End Class
    End Class
End Namespace

File: FluentMessageSenderCSharp/MessageSender.cs
using FluentMessageSenderCSharp.Interfaces;

namespace FluentMessageSenderCSharp
{

    public class MessageSender
    {

        private readonly IMessageSender _messageSender;
        public MessageSender(IMessageSender sender)
        {
            _messageSender = sender;
        }

        public SendExpression SendText(string text)
        {
            return new SendExpression(text, _messageSender);
        }

        public class SendExpression : IToExpression
        {

            private readonly string _text;
            private readonly IMessageSender _messageSender;

            private string _sender;
            public SendExpression(string text, IMessageSender messageSender)
            {
                _text = text;
                _messageSender = messageSender;
            }

            public IToExpression From(string sender)
            {
                _sender = sender;
                return this;
            }

            public void ToExpression(string receiver)
            {
                _messageSender.SendMessage(_text, _sender, receiver);
            }
            void IToExpression.To(string receiver)
            {
                ToExpression(receiver);
            }
        }
    }

 
}